lapis/connection.py

```diff
@@ -0,0 +1,115 @@
+        sitename
+        :param storage_element:
+        :return:
+        """
+        storage_element.remote_connection = self.remote_connection
+        try:
+            self.storages[storage_element.sitename].append(storage_element)
+        except KeyError:
+            self.storages[storage_element.sitename] = [storage_element]
+
+    async def _determine_inputfile_source(
+        self, requested_file: RequestedFile, dronesite: str, job_repr: str
+    ):
+        """
+        Collects NamedTuples containing the amount of data of the requested file
+        cached in a storage element and the storage element for all reachable storage
+        objects on the drone's site. The tuples are sorted by amount of cached data
+        and the storage object where the biggest part of the file is cached is
+        returned. If the file is not cached in any storage object the connection module
+        remote connection is returned.
+        :param requested_file:
+        :param dronesite:
+        :param job_repr:
+        :return:
+        """
+        provided_storages = self.storages.get(dronesite, None)
+        if provided_storages is not None:
+            look_up_list = []
+            for storage in provided_storages:
+                look_up_list.append(storage.look_up_file(requested_file, job_repr))
+            storage_list = sorted(
+                [entry async for entry in look_up_list],
+                key=lambda x: x[0],
+                reverse=True,
+            )
+            for entry in storage_list:
+                # TODO: check should better check that size is bigger than requested
+                if entry.cached_filesize > 0:
+                    return entry.storage
+        return self.remote_connection
+
+    async def stream_file(self, requested_file: RequestedFile, dronesite, job_repr):
+        """
+        Determines which storage object is used to provide the requested file and
+        startes the files transfer. For files transfered via remote connection a
+        potential cache decides whether to cache the file and handles the caching
+        process.
+        :param requested_file:
+        :param dronesite:
+        :param job_repr:
+        :return:
+        """
+        used_connection = await self._determine_inputfile_source(
+            requested_file, dronesite, job_repr
+        )
+
+        await sampling_required.put(used_connection)
+        if used_connection == self.remote_connection and self.storages.get(
+            dronesite, None
+        ):
+            try:
+                potential_cache = random.choice(self.storages[dronesite])
+                await potential_cache.apply_caching_decision(requested_file, job_repr)
+            except KeyError:
+                pass
+        print(f"now transfering {requested_file.filesize} from {used_connection}")
+        await used_connection.transfer(requested_file.filesize, job_repr)
```

Comment:
I like the idea you followed when you improved this but in the current state this only works in parts and we need to make a more general design decision in order to fix this. 
What I originally did was the following: 
If the file was not found the Connection objects `remote_connection` **Pipe** object was returned.  
If the file was found in some reachable cache by `self._determine_inputfile_source()` the corresponding **Storage** object was returned (not only the storage's Pipe object).
In order to be consistent when calling the actual file transfer in the line above this comment I implemented a `transfer()` method in the Storage class handling the actual usage of the storage's connection object and **updating the transferred files meta data**. In order to do this the method needs the the whole RequestedFile object and not only the filesize.
I agree that this implementation is not transparent and might be misleading. 

The current implementation does not work because if the requested file was found in a storage object the `transfer()` method is called passing `RequestedFile.filesize` instead of the expected `RequestedFile`. 

I can think of the following approaches to solve this problem: 

* separate the handling of the file transfer again and rename the storage objects `transfer()` method to show that there is a difference and avoid confusion
* update the files metadata somewhere else e.g. calling the `update()` method in the above function like the caching decision is called. I'm not sure whether this is a good way to handle this because I think that the stored files meta data handling should be done by the storage object and not the connection.
* I saw that you gave a remote connection to every storage object and I think that having different remote connections for different (groups of) caches will be necessary in order to simulate truely distributed systems and this might be a good opportunity to implement this. But as this is going to be a rather extensive change we should discuss our options tomorrow.

Reply:
Done. I introduced a new object ``RemoteStorage`` wrapping the ``Pipe``. So we have the same interface and can use the method as is.